### PR TITLE
[RNMobile] SegmentedControls correction

### PR DESCRIPTION
**Pull request by lukewalczak**
## Description

That PR is the result of further investigation mentioned in [temporary workaround](https://github.com/WordPress/gutenberg/pull/28698). It fixes the crash on Android platform occurs when segments are changed quickly.

Because the component _(custom indicator)_ located next to the `colors.map` is behind the condition and it’s a part of layout animation, the condition should be move to the state variable, which is solving the issue without rearranging components.

## How has this been tested?

1. Go to a post/page.
2. Add Buttons block.
3. Tap on the gear button.
4. Tap on `Background Color` options.
5. Switch several times between Solid and Gradient segments.

**Expect app mustn't crash**

## Types of changes
Bug fix

## Checklist:
- [X] My code is tested.
- [X] My code follows the WordPress code style. 
- [X] My code follows the accessibility standards. 
- [X] My code has proper inline documentation. 
- [X] I've included developer documentation if appropriate. 
- [X] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/components/src/color-palette/index.native.js [hunk 1]
```diff
@@ -51,6 +51,14 @@ function ColorPalette( {
 		'linear-gradient(360deg, rgba(0,0,255,.8) 0%, rgba(0,0,255,0) 70.71%)',
 	];
 
+	const [
+		shouldShowCustomIndicator,
+		setShouldShowCustomIndicator,
+	] = useState(
+		shouldShowCustomIndicatorOption &&
+			( ! isGradientSegment || isCustomGradientColor )
```
Comment (dratwas) [inline packages/components/src/color-palette/index.native.js:59]: Do we still need the `isCustomGradientColor` variable? Since  It is the same as  `isGradientColor` I suggest using the `isGradientColor` instead. It is also kinda confusing because you declare the `isCustomGradientColor` a few lines below :)

File: packages/components/src/color-palette/index.native.js [hunk 2]
```diff
@@ -208,45 +224,10 @@ function ColorPalette( {
...
-							) }
-						</View>
-					</TouchableWithoutFeedback>
-				</View>
-			) }
 			{ colors.map( ( color ) => {
 				const scaleValue = isSelected( color ) ? scaleInterpolation : 1;
 				return (
-					<View key={ `${ color }-${ isSelected( color ) }` }>
+					<View key={ color }>
```
Comment (dratwas) [inline packages/components/src/color-palette/index.native.js:230]: I'm just curious - why this change is needed? :) Or maybe do you know - why the `isSelected` was needed before?

**Review by dratwas — CHANGES_REQUESTED** [review]

File: packages/components/src/color-palette/index.native.js [hunk 3]
```diff
@@ -51,6 +51,14 @@ function ColorPalette( {
 		'linear-gradient(360deg, rgba(0,0,255,.8) 0%, rgba(0,0,255,0) 70.71%)',
 	];
 
+	const [
+		shouldShowCustomIndicator,
+		setShouldShowCustomIndicator,
+	] = useState(
+		shouldShowCustomIndicatorOption &&
+			( ! isGradientSegment || isCustomGradientColor )
+	);
```
Comment (dratwas) [inline packages/components/src/color-palette/index.native.js:60]:
```suggestion
```

File: packages/components/src/color-palette/index.native.js [hunk 4]
```diff
@@ -69,9 +77,17 @@ function ColorPalette( {
...
+	useEffect( () => {
+		setShouldShowCustomIndicator(
+			shouldShowCustomIndicatorOption &&
+				( ! isGradientSegment || isCustomGradientColor )
+		);
+	}, [
+		shouldShowCustomIndicatorOption,
+		isGradientSegment,
+		isCustomGradientColor,
+	] );
```
Comment (dratwas) [inline packages/components/src/color-palette/index.native.js:90]:
```suggestion
	const shouldShowCustomIndicator = useMemo( () => {
		return (
			shouldShowCustomIndicatorOption &&
			( ! isGradientSegment || isCustomGradientColor )
		);
	}, [
		shouldShowCustomIndicatorOption,
		isGradientSegment,
		isCustomGradientColor,
	] );
```

**Review by dratwas — COMMENTED** [review]
The fix works perfectly, however, I left one suggestion. Could you please check if it works with using memo instead of state?

**Comment by lukewalczak** [comment]
@dratwas thanks for a suggestion, however using memo crashes the app again. As I mentioned in the description variable called `shouldShowCustomIndicator` should be located in the state.

**Review by dratwas — APPROVED** [review]
